File: scripts/load-lessons.js
#!/usr/bin/env node

/**
 * Firebase Lesson Data Loader
 *
 * This script loads sample lessons from content/lessons.json into Firestore.
 *
 * Usage:
 *   node scripts/load-lessons.js
 *
 * Requirements:
 *   - Firebase Admin SDK initialized with service account
 *   - Service account JSON in backend/nodejs/config/firebase-key.json
 *   - Firestore database created in Firebase Console
 */

const admin = require('firebase-admin');
const path = require('path');
const fs = require('fs');

// Get service account path
const serviceAccountPath = path.join(__dirname, '../backend/nodejs/config/firebase-key.json');

// Check if service account exists
if (!fs.existsSync(serviceAccountPath)) {
  console.error('❌ Error: Service account file not found at:', serviceAccountPath);
  console.error('\nPlease:');
  console.error('1. Go to Firebase Console → Project Settings → Service Accounts');
  console.error('2. Click "Generate New Private Key"');
  console.error('3. Save the JSON file to:', serviceAccountPath);
  process.exit(1);
}

// Initialize Firebase Admin
try {
  const serviceAccount = require(serviceAccountPath);

  admin.initializeApp({
    credential: admin.credential.cert(serviceAccount)
  });

  console.log('✓ Firebase Admin SDK initialized');
} catch (error) {
  console.error('❌ Error initializing Firebase:', error.message);
  process.exit(1);
}

// Load lessons data
const lessonsPath = path.join(__dirname, '../content/lessons.json');
let lessonsData;

try {
  lessonsData = require(lessonsPath);
  console.log(`✓ Loaded ${lessonsData.length} lessons from content/lessons.json`);
} catch (error) {
  console.error('❌ Error loading lessons:', error.message);
  process.exit(1);
}

// Get Firestore instance
const db = admin.firestore();

/**
 * Load lessons into Firestore
 */
async function loadLessons() {
  console.log('\n📚 Loading lessons into Firestore...\n');

  let successCount = 0;
  let errorCount = 0;

  for (const lesson of lessonsData) {
    try {
      // Validate lesson has required fields
      if (!lesson.id || !lesson.title) {
        throw new Error('Lesson missing id or title');
      }

      // Set the lesson document
      await db.collection('lessons').doc(lesson.id).set(lesson);

      console.log(`✓ Loaded: "${lesson.title}"`);
      successCount++;
    } catch (error) {
      console.error(`✗ Error loading lesson: ${error.message}`);
      errorCount++;
    }
  }

  // Summary
  console.log('\n' + '='.repeat(50));
  console.log(`Summary:`);
  console.log(`✓ Successfully loaded: ${successCount} lessons`);
  if (errorCount > 0) {
    console.log(`✗ Errors: ${errorCount} lessons`);
  }
  console.log('='.repeat(50) + '\n');

  // Verify data
  console.log('Verifying data in Firestore...\n');

  try {
    const snapshot = await db.collection('lessons').get();
    console.log(`✓ Total lessons in Firestore: ${snapshot.size}`);

    if (snapshot.size > 0) {
      console.log('\nLessons in database:');
      snapshot.forEach(doc => {
        const lesson = doc.data();
        console.log(`  • ${lesson.title} (${lesson.category})`);
      });
    }
  } catch (error) {
    console.error('✗ Error verifying data:', error.message);
  }

  // Exit
  if (errorCount === 0) {
    console.log('\n✓ All lessons loaded successfully!');
    process.exit(0);
  } else {
    console.log(`\n✗ ${errorCount} errors occurred.`);
    process.exit(1);
  }
}

// Run
loadLessons().catch(error => {
  console.error('❌ Fatal error:', error.message);
  process.exit(1);
});

File: scripts/load_lessons.py
#!/usr/bin/env python
"""
Firebase Lesson Data Loader

This script loads sample lessons from content/lessons.json into Firestore.

Usage:
    python scripts/load_lessons.py

Requirements:
    - Firebase Admin SDK installed
    - Service account JSON in backend/django/config/firebase-key.json
    - Firestore database created in Firebase Console
"""

import json
import os
import sys
from pathlib import Path

import firebase_admin
from firebase_admin import credentials, firestore

# Get base directory
BASE_DIR = Path(__file__).resolve().parent.parent

# Service account path
SERVICE_ACCOUNT_PATH = BASE_DIR / 'backend/django/config/firebase-key.json'

# Check if service account exists
if not SERVICE_ACCOUNT_PATH.exists():
    print('❌ Error: Service account file not found at:', SERVICE_ACCOUNT_PATH)
    print('\nPlease:')
    print('1. Go to Firebase Console → Project Settings → Service Accounts')
    print('2. Click "Generate New Private Key"')
    print('3. Save the JSON file to:', SERVICE_ACCOUNT_PATH)
    sys.exit(1)

# Initialize Firebase Admin
try:
    cred = credentials.Certificate(str(SERVICE_ACCOUNT_PATH))
    firebase_admin.initialize_app(cred)
    print('✓ Firebase Admin SDK initialized')
except Exception as error:
    print(f'❌ Error initializing Firebase: {error}')
    sys.exit(1)

# Load lessons data
lessons_path = BASE_DIR / 'content/lessons.json'

try:
    with open(lessons_path, 'r') as f:
        lessons_data = json.load(f)
    print(f'✓ Loaded {len(lessons_data)} lessons from content/lessons.json')
except Exception as error:
    print(f'❌ Error loading lessons: {error}')
    sys.exit(1)

# Get Firestore instance
db = firestore.client()


def load_lessons():
    """Load lessons into Firestore"""
    print('\n📚 Loading lessons into Firestore...\n')

    success_count = 0
    error_count = 0

    for lesson in lessons_data:
        try:
            # Validate lesson has required fields
            if not lesson.get('id') or not lesson.get('title'):
                raise ValueError('Lesson missing id or title')

            # Set the lesson document
            db.collection('lessons').document(lesson['id']).set(lesson)

            print(f'✓ Loaded: "{lesson["title"]}"')
            success_count += 1
        except Exception as error:
            print(f'✗ Error loading lesson: {error}')
            error_count += 1

    # Summary
    print('\n' + '=' * 50)
    print('Summary:')
    print(f'✓ Successfully loaded: {success_count} lessons')
    if error_count > 0:
        print(f'✗ Errors: {error_count} lessons')
    print('=' * 50 + '\n')

    # Verify data
    print('Verifying data in Firestore...\n')

    try:
        docs = db.collection('lessons').get()
        doc_list = list(docs)
        print(f'✓ Total lessons in Firestore: {len(doc_list)}')

        if doc_list:
            print('\nLessons in database:')
            for doc in doc_list:
                lesson = doc.to_dict()
                print(f'  • {lesson["title"]} ({lesson["category"]})')
    except Exception as error:
        print(f'✗ Error verifying data: {error}')

    # Exit
    if error_count == 0:
        print('\n✓ All lessons loaded successfully!')
        sys.exit(0)
    else:
        print(f'\n✗ {error_count} errors occurred.')
        sys.exit(1)


if __name__ == '__main__':
    try:
        load_lessons()
    except Exception as error:
        print(f'❌ Fatal error: {error}')
        sys.exit(1)
